### PR TITLE
Build Metis for Windows

### PR DESCRIPTION
**Pull request by basnijholt**

### CODE REVIEW

**Comment by jakirkham** [comment]
FYI PR ( https://github.com/conda-forge/metis-feedstock/pull/8 ) will fix the lint issue. Once merged, the lint error should be resolved in this PR as well.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -11,25 +11,41 @@ source:
...
+  features:
+    - vc9  # [win and py27]
+    - vc10  # [win and py34]
+    - vc14  # [win and py35]
 
 requirements:
   build:
     - cmake
     - toolchain
+    - python  # [win]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: Adding to the `build` and `run` section is probably good practice. One day we may drop the `python` dependency and use only the `vc`s. (But for now you still need both or the re-rendering will mess up the build matrix.)

```yaml
   - vc 9  # [win and py27]
    - vc 10  # [win and py34]
    - vc 14  # [win and py35]
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -11,25 +11,41 @@ source:
...
 
 test:
+  requires:
+    - python 2.7.*  # [win and py27]
+    - python 3.4.*  # [win and py34]
+    - python 3.5.*  # [win and py35]
```
Comment (ocefpaf) [inline recipe/meta.yaml]: If you add the `vc`s from above you can remove this.

Comment (basnijholt) [inline recipe/meta.yaml]: Thanks, removed in 9dc9415.

Comment (ocefpaf) [inline recipe/meta.yaml]: You need to do what I mention in https://github.com/conda-forge/metis-feedstock/pull/7#discussion_r87870092 for that to work properly. Sorry if I was not clear before.

**Comment by basnijholt** [comment]
@isuruf, I removed it because it's now in the build and run requirements, @ocefpaf explained it some comments above.

**Comment by isuruf** [comment]
You can try NMake again, because now cmake line is fixed.

File: recipe/bld.bat [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+CD build\windows
+
+cmake ^
+    -G "NMake Makefiles"                     ^
+    -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX%  ^
+    -DCMAKE_INSTALL_BINDIR=%LIBRARY_BIN%     ^
+    -DCMAKE_INSTALL_LIBDIR=%LIBRARY_LIB%     ^
+    -DCMAKE_INSTALL_INCLUDEDIR=%LIBRARY_INC% ^
+    -DCMAKE_PREFIX_PATH=%LIBRARY_PREFIX%     ^
+    -DCMAKE_BUILD_TYPE=Release               ^
```
Comment (isuruf) [inline recipe/bld.bat]: This is building a static library. @jakirkham, which is preferred? static or shared?

File: recipe/bld.bat [hunk 4]
```diff
@@ -0,0 +1,28 @@
+MKDIR build\windows
+CD build\windows
+
+cmake ^
+    -G "NMake Makefiles"                     ^
+    -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX%  ^
+    -DCMAKE_INSTALL_BINDIR=%LIBRARY_BIN%     ^
+    -DCMAKE_INSTALL_LIBDIR=%LIBRARY_LIB%     ^
+    -DCMAKE_INSTALL_INCLUDEDIR=%LIBRARY_INC% ^
```
Comment (isuruf) [inline recipe/bld.bat]: Above 3 lines are unnecessary

File: recipe/bld.bat [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+dir include
+dir Release
+
+copy libmetis\metis.lib %LIBRARY_BIN%
+copy programs\cmpfillin.exe %LIBRARY_LIB%
+copy programs\gpmetis.exe %LIBRARY_LIB%
+copy programs\graphchk.exe %LIBRARY_LIB%
+copy programs\m2gmetis.exe %LIBRARY_LIB%
+copy programs\ndmetis.exe %LIBRARY_LIB%
+copy programs\mpmetis.exe %LIBRARY_LIB%
```
Comment (isuruf) [inline recipe/bld.bat]: `.h` is not copied over.

File: recipe/bld.bat [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+cmake --build . --config Release
+
+dir
+dir libmetis
+dir programs
+dir include
+dir Release
+
+copy libmetis\metis.lib %LIBRARY_BIN%
```
Comment (isuruf) [inline recipe/bld.bat]: `.lib` is usually placed in `lib` and `.dll` in `bin`

File: recipe/meta.yaml [hunk 7]
```diff
@@ -6,34 +6,55 @@ package:
...
-    - m2gmetis -h
-    - test -f $PREFIX/include/metis.h
+    - graphchk  # [unix]
+    - cmpfillin -h  # [unix]
+    - mpmetis -h  # [unix]
+    - gpmetis -h  # [unix]
+    - ndmetis -h  # [unix]
+    - gpmetis -h  # [unix]
+    - m2gmetis -h  # [unix]
+    - test -f $PREFIX/include/metis.h  # [unix]
```
Comment (isuruf) [inline recipe/meta.yaml]: Library is not installed in `unix`?

Comment (basnijholt) [inline recipe/meta.yaml]: idk, @minrk added this line.

File: recipe/bld.bat (hunk 4 above)
Comment (basnijholt) [inline recipe/bld.bat]: Removed in f2b8f8e.

File: recipe/bld.bat (hunk 6 above)
Comment (basnijholt) [inline recipe/bld.bat]: I corrected it in f2b8f8e and also put `.exe` in `bin`.

File: recipe/bld.bat (hunk 5 above)
Comment (basnijholt) [inline recipe/bld.bat]: Fixed in f2b8f8e.

File: recipe/bld.bat [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+copy programs\cmpfillin.exe %LIBRARY_BIN%
+copy programs\gpmetis.exe %LIBRARY_BIN%
+copy programs\graphchk.exe %LIBRARY_BIN%
+copy programs\m2gmetis.exe %LIBRARY_BIN%
+copy programs\ndmetis.exe %LIBRARY_BIN%
+copy programs\mpmetis.exe %LIBRARY_BIN%
+copy ..\..\include\metis.h %LIBRARY_INC%
+
+CD programs
+mpmetis.exe ..\..\..\graphs\metis.mesh 10
```
Comment (isuruf) [inline recipe/bld.bat]: Add `if errorlevel 1 exit 1` after each line to check for errors

Comment (isuruf) [inline recipe/bld.bat]: Sorry if I made this unclear. `if errorlevel 1 exit 1` needs to be on a new line after each exe call.

Comment (basnijholt) [inline recipe/bld.bat]: Fixed in 676e9d8.

File: recipe/meta.yaml [hunk 9]
```diff
@@ -43,14 +43,14 @@ test:
...
-    - if not exist "%LIBRARY_LIB%\\gpmetis.exe" exit 1  # [win]
-    - if not exist "%LIBRARY_LIB%\\m2gmetis.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_LIB%\\metis.lib" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\graphchk.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\cmpfillin.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\mpmetis.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\gpmetis.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\ndmetis.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\gpmetis.exe" exit 1  # [win]
+    - if not exist "%LIBRARY_BIN%\\m2gmetis.exe" exit 1  # [win]
```
Comment (isuruf) [inline recipe/meta.yaml]: :+1:

Comment (isuruf) [inline recipe/meta.yaml]: Can you check for the header in windows as well?

File: recipe/bld.bat (hunk 3 above)
Comment (jakirkham) [inline recipe/bld.bat]: We had a lot of trouble trying to get `metis` to build a shared library even on Unix. So ultimately we decided it was ok to just build a static library on Unix. I think it is reasonable for it to do the same on Windows.

**Comment by jakirkham** [comment]
Thanks very much for helping get this in good shape @isuruf . Do you use `metis` at all? Would you have any interest in helping maintain it?

**Comment by isuruf** [comment]
I still have no clue what metis is. :smile:

File: recipe/meta.yaml (hunk 7 above)
Comment (basnijholt) [inline recipe/meta.yaml]: (Correctly) did it in 2fad0cf.

**Comment by jakirkham** [comment]
I'll take that as a "no". 😄

It's a [graph partitioning]( https://en.wikipedia.org/wiki/Graph_partition ) library. They list some other things it can do, but they all reduce to that.

File: recipe/meta.yaml [hunk 10]
```diff
@@ -6,34 +6,58 @@ package:
...
+    - graphchk  # [unix]
+    - cmpfillin -h  # [unix]
+    - mpmetis -h  # [unix]
+    - gpmetis -h  # [unix]
+    - ndmetis -h  # [unix]
+    - gpmetis -h  # [unix]
+    - m2gmetis -h  # [unix]
+    - test -f $PREFIX/include/metis.h  # [unix]
+    - test -f $PREFIX/lib/libmetis.dylib  # [osx]
+    - test -f $PREFIX/lib/libmetis.so  # [linux]
```
Comment (basnijholt) [inline recipe/meta.yaml]: Lucky I guess, haha.

Comment (basnijholt) [inline recipe/meta.yaml]: Make Windows libs shared too?

Comment (isuruf) [inline recipe/meta.yaml]: Add `-DBUILD_SHARED_LIBS=yes` to cmake and copy both `.dll` and `.lib`

Comment (basnijholt) [inline recipe/meta.yaml]: Done in e218b0f. However, there are no `.dll`s created.

Comment (isuruf) [inline recipe/meta.yaml]: Metis is overriding default cmake behaviour. Try `-DSHARED=yes`

Comment (basnijholt) [inline recipe/meta.yaml]: It seems to not really work.

Comment (isuruf) [inline recipe/meta.yaml]: From the logs, it looks like the `.lib` is not generated. Usually this happens when the functions are not exported. I looked at the header file and it seems you need to define `_WINDLL`.
Try something like, `set CFLAGS="/D_WINDLL" at the top of bld.bat before invoking cmake.

**Comment by isuruf** [comment]
executables are using functions not defined in `metis.h` which causes the error.
Try adding `-DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=yes` to cmake.
If that doesn't work, I guess static library will have to do.

**Comment by basnijholt** [comment]
Unfortunately, also doesn't do the trick. Shall I revert to static lib compilation?

**Comment by isuruf** [comment]
You can create the executables using a static build and build the dll using a shared build (`dll` creation is fine, creating the executables is the problem), but that's kind of hacky.
So, reverting to static lib is the best option.

**Comment by basnijholt** [comment]
I reverted it and merged, @isuruf thanks a lot for helping me on this one! 👍

**Comment by jakirkham** [comment]
Does this work though? 😕 Seems like the CIs didn't complete. Did you test on your fork or something?

**Comment by basnijholt** [comment]
Hmmm, I just reverted to a commit that did pass I thought... Should I create another PR with correct script?

**Comment by basnijholt** [comment]
It worked [here](https://ci.appveyor.com/project/basnijholt/metis-feedstock) too.

**Comment by basnijholt** [comment]
Actually, [CircleCI](https://circleci.com/gh/conda-forge/metis-feedstock), [Travis](https://travis-ci.org/conda-forge/metis-feedstock), and [AppVeyor](https://ci.appveyor.com/project/conda-forge/metis-feedstock/branch/master) all passed.

**Pull request merged**